ops/manifests/manifest.py

```diff
@@ -0,0 +1,329 @@
+        """All unique component resources.
+
+        Order is guaranteed to be:
+        * Addition Manipulations
+        * Manifest files contents
+        """
+        result: Dict[HashableResource, None] = OrderedDict()
+        ver = self.current_release
+
+        # Generated additions
+        for manipulate in self.manipulations:
+            if isinstance(manipulate, Addition):
+                obj = manipulate()
+                if not obj:
+                    continue
+                result[HashableResource(obj)] = None
+
+        # From static manifests
+        for manifest in (self.manifest_path / ver).glob("*.yaml"):
+            for obj in self._safe_load(manifest):
+                result[HashableResource(obj)] = None
+
+        return result.keys()
+
+    @lru_cache()
+    def _safe_load(self, filepath: Path) -> List[AnyResource]:
+        """Read manifest file and parse its content into lightkube objects.
+
+        Lightkube can't properly read manifest files which contain List kinds.
+        """
+        content = filepath.read_text()
+        return [
+            codecs.from_dict(item)  # Map to lightkube resources
+            for rsc in yaml.safe_load_all(content)  # load content from file
+            if rsc  # ignore empty objects
+            for item in (rsc["items"] if rsc["kind"] == "List" else [rsc])
+        ]
+
+    def status(self) -> FrozenSet[HashableResource]:
+        """Returns all installed objects which have a `.status.conditions` attribute."""
+        return frozenset(_ for _ in self.installed_resources() if _.status_conditions)
+
+    def installed_resources(self) -> FrozenSet[HashableResource]:
+        """All currently installed resources expected by this manifest."""
+        result: Dict[HashableResource, None] = OrderedDict()
+        for obj in self.resources:
+            try:
+                next_rsc = self.client.get(
+                    type(obj.resource),
+                    obj.name,
+                    namespace=obj.namespace,
+                )
+            except ApiError:
+                continue
+            result[HashableResource(next_rsc)] = None
+        return frozenset(result.keys())
+
+    def labelled_resources(self) -> FrozenSet[HashableResource]:
+        """Any resource ever installed and labeled by this class."""
+
+        NamespaceKind = namedtuple("NamespaceKind", "namespace, kind")
+        ns_kinds = set(
+            NamespaceKind(obj.namespace, type(obj.resource)) for obj in self.resources
+        )
+
+        return frozenset(
+            HashableResource(rsc)
+            for ns_kind in ns_kinds
+            for rsc in self.client.list(
+                ns_kind.kind,
+                namespace=ns_kind.namespace,
+                labels={
+                    "juju.io/application": self.app_name,
+                    "juju.io/manifest": self.name,
+                },
+            )
+        )
+
+    def apply_manifests(self):
```

Comment:
Could there be a time when youd only want to apply certain manifest files but not others?

Reply:
hmm... i'm thinking there could be situations where someone would want to "restart" certain things (ie deployments). 

i'd envisioned something like :

```python
   manifest.delete_resource(thing)  # delete something to reapply
   manifest.apply_manifests()         # apply everything again
```

but i can see what you're asking for

Reply:
yeah I was thinking of a case where say all the one-time patching and additions have been applied on install, but then maybe config changes or something and you just need to modify a deployment and not update EVERYTHING

Reply:
hows this @stonepreston